tooling/nargo_cli/src/cli/fuzz_cmd.rs

```diff
@@ -0,0 +1,489 @@
+    FuzzExecutionConfig, FuzzFolderConfig,
+};
+use nargo_toml::PackageSelection;
+use noirc_abi::input_parser::{json::serialize_to_json, Format};
+use noirc_driver::{check_crate, CompileOptions};
+use noirc_frontend::hir::{FunctionNameMatch, ParsedFiles};
+use rayon::prelude::{ParallelBridge, ParallelIterator};
+use termcolor::{Color, ColorChoice, ColorSpec, StandardStream, WriteColor};
+
+use crate::{cli::check_cmd::check_crate_and_report_errors, errors::CliError};
+
+use super::{fs::inputs::write_inputs_to_file, LockType, PackageOptions, WorkspaceCommand};
+
+/// Run the tests for this program
+#[derive(Debug, Clone, Args)]
+#[clap(visible_alias = "f")]
+pub(crate) struct FuzzCommand {
+    /// If given, only the fuzzing harnesses with names containing this string will be run
+    fuzzing_harness_name: Option<String>,
+
+    /// If given, load/store fuzzer corpus from this folder
+    #[arg(long)]
+    corpus_dir: Option<String>,
+
+    /// If given, perform corpus minimization instead of fuzzing and store results in the given folder
+    #[arg(long)]
+    minimized_corpus_dir: Option<String>,
+
+    /// If given, store the failing input in the given folder
+    #[arg(long)]
+    fuzzing_failure_dir: Option<String>,
+    /// List all available harnesses that match the name
+    #[clap(long)]
+    list_all: bool,
+
+    /// Display output of `println` statements
+    #[arg(long)]
+    show_output: bool,
+
+    /// The number of threads to use for fuzzing
+    #[arg(long, default_value = "1")]
+    num_threads: usize,
+
+    /// Only run harnesses that match exactly
+    #[clap(long)]
+    exact: bool,
+
+    #[clap(flatten)]
+    pub(super) package_options: PackageOptions,
+
+    #[clap(flatten)]
+    compile_options: CompileOptions,
+
+    /// JSON RPC url to solve oracle calls
+    #[clap(long)]
+    oracle_resolver: Option<String>,
+
+    /// Maximum time in seconds to spend fuzzing (default: no timeout)
+    #[arg(long)]
+    timeout: Option<u64>,
+}
+impl WorkspaceCommand for FuzzCommand {
+    fn package_selection(&self) -> PackageSelection {
+        self.package_options.package_selection()
+    }
+    fn lock_type(&self) -> LockType {
+        // Reads the code to compile fuzzing harnesses in memory, but doesn't save artifacts.
+        LockType::None
+    }
+}
+pub(crate) fn run(args: FuzzCommand, workspace: Workspace) -> Result<(), CliError> {
+    let mut file_manager = workspace.new_file_manager();
+    insert_all_files_for_workspace_into_file_manager(&workspace, &mut file_manager);
+    let parsed_files = parse_all(&file_manager);
+
+    let pattern = match &args.fuzzing_harness_name {
+        Some(name) => {
+            let names = vec![name.to_string()];
+            if args.exact {
+                FunctionNameMatch::Exact(names)
+            } else {
+                FunctionNameMatch::Contains(names)
+            }
+        }
+        None => FunctionNameMatch::Anything,
+    };
+
+    if args.list_all {
```

Comment:
The contents of this `if` could be moved to a helper function.

Reply:
Done